nest-forms-backend/src/utils/subservices/sharepoint.subservice.ts

```diff
@@ -0,0 +1,357 @@
+  ) {
+    this.logger = new Logger('SharepointSubservice')
+
+    if (
+      !process.env.SHAREPOINT_TENANT_ID ||
+      !process.env.SHAREPOINT_CLIENT_ID ||
+      !process.env.SHAREPOINT_CLIENT_SECRET ||
+      !process.env.SHAREPOINT_DOMAIN ||
+      !process.env.SHAREPOINT_URL
+    ) {
+      throw new Error(
+        'Missing Sharepoint .env values, one of: SHAREPOINT_TENANT_ID, SHAREPOINT_CLIENT_ID, SHAREPOINT_CLIENT_SECRET, SHAREPOINT_DOMAIN, SHAREPOINT_URL',
+      )
+    }
+  }
+
+  @Process()
+  async transcode(job: Job<{ formId: string }>): Promise<void> {
+    await this.postNewRecord(job.data.formId)
+  }
+
+  @OnQueueFailed()
+  handler(job: Job<{ formId: string }>, err: Error): void {
+    alertError(
+      `Sending form ${job.data.formId} to Sharepoint has failed. Number of tries for this form so far: ${job.attemptsMade}`,
+      this.logger,
+      JSON.stringify(err),
+    )
+
+    this.prismaService.forms
+      .update({
+        where: {
+          id: job.data.formId,
+        },
+        data: {
+          error: FormError.SHAREPOINT_SEND_ERROR,
+        },
+      })
+      .catch((error) => {
+        alertError(
+          `Setting form error with id ${job.data.formId} to POWERAPPS_SEND_ERROR failed.`,
+          this.logger,
+          JSON.stringify(error),
+        )
+      })
+  }
+
+  public async postNewRecord(formId: string): Promise<void> {
+    const form = await this.prismaService.forms.findUnique({
+      where: {
+        id: formId,
+        archived: false,
+      },
+    })
+    if (form === null) {
+      throw this.throwerErrorGuard.NotFoundException(
+        FormsErrorsEnum.FORM_NOT_FOUND_ERROR,
+        FormsErrorsResponseEnum.FORM_NOT_FOUND_ERROR,
+      )
+    }
+
+    const formDefinition = getFormDefinitionBySlug(form.formDefinitionSlug)
+    if (!formDefinition) {
+      throw this.throwerErrorGuard.NotFoundException(
+        FormsErrorsEnum.FORM_DEFINITION_NOT_FOUND,
+        `${FormsErrorsResponseEnum.FORM_DEFINITION_NOT_FOUND} ${form.formDefinitionSlug}`,
+      )
+    }
+
+    if (
+      formDefinition.type !== FormDefinitionType.SlovenskoSkGeneric ||
+      !formDefinition.sharepointData
+    ) {
+      throw this.throwerErrorGuard.UnprocessableEntityException(
+        SharepointErrorsEnum.SHAREPOINT_DATA_NOT_PROVIDED,
+        `${SharepointErrorsResponseEnum.SHAREPOINT_DATA_NOT_PROVIDED} Form id: ${form.id}.`,
+      )
+    }
+
+    const accessToken = await this.getAccessToken()
+    const { sharepointData, schemas } = formDefinition
+    const formTitle = this.getTitle(form, formDefinition)
+    const fields = await this.mapColumnsToFields(
+      Object.keys(sharepointData.columnMap),
+      accessToken,
+      sharepointData.databaseName,
+    )
+    const jsonDataExtraDataOmitted = omitExtraData(
+      schemas.schema,
+      form.formDataJson as GenericObjectType,
+    )
+    const valuesForFields = getValuesForFields(
+      sharepointData,
```

Comment:
Does it make sense to move `omitExtraData` to be called inside `getValuesForFields`?

Reply:
Maybe not, the implementation will be tied to form definitions, which makes it hard to create a simple unit tests.

Reply:
Yes, I would call it once here and then use the result everywhere